internal/cloud/awscloud/secure-instance.go

```diff
@@ -0,0 +1,396 @@
+package awscloud
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/sirupsen/logrus"
+)
+
+type SecureInstance struct {
+	FleetID  string
+	SGID     string
+	LTID     string
+	Instance *ec2.Instance
+}
+
+// Runs an instance with a security group that only allows traffic to
+// the host. Will replace resources if they already exists.
+func (a *AWS) RunSecureInstance(iamProfile string) (*SecureInstance, error) {
+	identity, err := a.ec2metadata.GetInstanceIdentityDocument()
+	if err != nil {
+		logrus.Errorf("Error getting the identity document, %s", err)
+		return nil, err
+	}
+
+	descrInstancesOutput, err := a.ec2.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{
+			aws.String(identity.InstanceID),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(descrInstancesOutput.Reservations) != 1 || len(descrInstancesOutput.Reservations[0].Instances) != 1 {
+		return nil, fmt.Errorf("Expected exactly one instance")
+	}
+	vpcID := *descrInstancesOutput.Reservations[0].Instances[0].VpcId
+	imageID := *descrInstancesOutput.Reservations[0].Instances[0].ImageId
+
+	secureInstance := &SecureInstance{}
+	defer func() {
+		if secureInstance.Instance == nil {
+			err := a.deleteFleetIfExists(secureInstance)
+			if err != nil {
+				logrus.Errorf("Deleting fleet unsuccessful: %v", err)
+			}
+
+			err = a.deleteLTIfExists(secureInstance)
+			if err != nil {
+				logrus.Errorf("Error deleting  launch template %s: %v", secureInstance.LTID, err)
+			}
+
+			err = a.deleteSGIfExists(secureInstance)
+			if err != nil {
+				logrus.Errorf("Error deleting security group %s: %v", secureInstance.SGID, err)
+			}
+		}
+	}()
+
+	sgID, err := a.createOrReplaceSG(identity.InstanceID, identity.PrivateIP, vpcID)
+	if sgID != "" {
+		secureInstance.SGID = sgID
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	ltID, err := a.createOrReplaceLT(identity.InstanceID, imageID, sgID, iamProfile)
+	if ltID != "" {
+		secureInstance.LTID = ltID
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	createFleetOutput, err := a.ec2.CreateFleet(&ec2.CreateFleetInput{
+		LaunchTemplateConfigs: []*ec2.FleetLaunchTemplateConfigRequest{
```

Comment:
I think this is just a feature of the aws sdk :/ I'm not sure if it would be that much better considering it's already in the private `createOrReplaceLT` function.